File: output/objc/PersonLenses.m
#import "PersonLenses.h"

@implementation PersonLenses

+ (Lens<Person*, NSString*>*)firstName
{
    return [Lens lensWithGet:^NSString*(Person* whole){
        return whole.firstName;
    } set:^Person*(Person* whole, NSString* part) {
        return [whole personBySettingFirstName:part];
    }];
}
+ (Lens<Person*, NSString*>*)lastName
{
    return [Lens lensWithGet:^NSString*(Person* whole){
        return whole.lastName;
    } set:^Person*(Person* whole, NSString* part) {
        return [whole personBySettingLastName:part];
    }];
}
+ (Lens<Person*, NSString*>*)nickName
{
    return [Lens lensWithGet:^NSString*(Person* whole){
        return whole.nickName;
    } set:^Person*(Person* whole, NSString* part) {
        return [whole personBySettingNickName:part];
    }];
}
+ (Lens<Person*, NSNumber*>*)age
{
    return [Lens lensWithGet:^NSNumber*(Person* whole){
        return @(whole.age);
    } set:^Person*(Person* whole, NSNumber* part) {
        return [whole personBySettingAge:part.integerValue];
    }];
}
+ (Lens<Person*, NSNumber*>*)canOrder
{
    return [Lens lensWithGet:^NSNumber*(Person* whole){
        return @(whole.canOrder);
    } set:^Person*(Person* whole, NSNumber* part) {
        return [whole personBySettingCanOrder:part.boolValue];
    }];
}
+ (Lens<Person*, Address*>*)address
{
    return [Lens lensWithGet:^Address*(Person* whole){
        return whole.address;
    } set:^Person*(Person* whole, Address* part) {
        return [whole personBySettingAddress:part];
    }];
}

@end

@implementation BoundLensToPerson

- (BoundLens*)firstName 
{
    return [BoundLens lensWithParent:self sublens:[PersonLenses firstName]];
}
- (BoundLens*)lastName 
{
    return [BoundLens lensWithParent:self sublens:[PersonLenses lastName]];
}
- (BoundLens*)nickName 
{
    return [BoundLens lensWithParent:self sublens:[PersonLenses nickName]];
}
- (BoundLens*)age 
{
    return [BoundLens lensWithParent:self sublens:[PersonLenses age]];
}
- (BoundLens*)canOrder 
{
    return [BoundLens lensWithParent:self sublens:[PersonLenses canOrder]];
}
- (BoundLensToAddress *)address 
{
    return [BoundLensToAddress lensWithParent:self sublens:[PersonLenses address]];
}

@end

@implementation Person (Lens)

- (BoundLensToPerson*)lens
{
    return [BoundLensToPerson lensWithInstance:self lens:[Lens identityLens]];
}

@end

File: output/swift/PersonLenses.swift
import Foundation

extension Person {
    struct Lenses {
        static let firstName: Lens<Person, String> = Lens(
            get: { $0.firstName },
            set: { (person, firstName) in
                var builder = PersonBuilder(existingPerson: person)
                return builder.withFirstName(firstName).build()
            }
        )
        static let lastName: Lens<Person, String> = Lens(
            get: { $0.lastName },
            set: { (person, lastName) in
                var builder = PersonBuilder(existingPerson: person)
                return builder.withLastName(lastName).build()
            }
        )
        static let nickName: Lens<Person, String> = Lens(
            get: { $0.nickName },
            set: { (person, nickName) in
                var builder = PersonBuilder(existingPerson: person)
                return builder.withNickName(nickName).build()
            }
        )
        static let age: Lens<Person, Int> = Lens(
            get: { $0.age },
            set: { (person, age) in
                var builder = PersonBuilder(existingPerson: person)
                return builder.withAge(age).build()
            }
        )
        static let address: Lens<Person, Address> = Lens(
            get: { $0.address },
            set: { (person, address) in
                var builder = PersonBuilder(existingPerson: person)
                return builder.withAddress(address).build()
            }
        )
    }
}

struct BoundLensToPerson<Whole>: BoundLensType {
    typealias Part = Person
    let storage: BoundLensStorage<Whole, Part>
    
    var firstName: BoundLens<Whole, String> {
        return BoundLens<Whole, String>(parent: self, sublens: Person.Lenses.firstName)
    }

    var lastName: BoundLens<Whole, String> {
        return BoundLens<Whole, String>(parent: self, sublens: Person.Lenses.lastName)
    }

    var nickName: BoundLens<Whole, String> {
        return BoundLens<Whole, String>(parent: self, sublens: Person.Lenses.nickName)
    }

    var age: BoundLens<Whole, Int> {
        return BoundLens<Whole, Int>(parent: self, sublens: Person.Lenses.age)
    }

    var address: BoundLensToAddress<Whole> {
        return BoundLensToAddress<Whole>(parent: self, sublens: Person.Lenses.address)
    }
}

extension Person {
    var lens: BoundLensToPerson<Person> {
        return BoundLensToPerson<Person>(instance: self, lens: createIdentityLens())
    }
}
